### ISSUE TITLE
Docs for running WebPageTest headlessly

### ISSUE THREAD

**Issue by iDVB** [WPO-Foundation/wptagent#179]
Is there a way to run WebPageTest (Docker) "tests" headlessly?
We'd love to be able to add a WebPageTest to our CI pipeline without having to continuosly run server and agent containers but instead just fire them up as needed.

We'd like these to run headlessly as well.... so that the results of the tests can be output or boiled down to a single score that can affect our pipeline. 

We're going this successfully now with Lighthouse (Docker).

**Comment by pmeenan** [comment]
Do you mean on-demand more than headlessly?  Both the server and (linux) agent run headless but there is no push-model like with Lighthouse.  You'd still have to spin up a server and agent to run the test (and kill them after if you wanted to).

Both webpagetest and wptagent (the agent) have docker containers available though there are some caveats around the agent docker container because it needs access to do system-wide traffic-shaping.

**Comment by iDVB** [comment]
@pmeenan Hmmm.... ya ... I mean both. Ondemand and headless.
I'd like to have a config file I can pass into the container when it loads that just tests the site link specified using the test specified and then provide the results and shut down.

I do not want to have to manually click around the server UI to run a test.

If this is possible.... is there any docs detailing out how this can be done with containers? I've taken a look at both server and agent containers... and can see that they both have Dockerfiles... and they detail out the settings you can pass.... but I don't see any settings for.... "url to test" or "test to run?" kind of thing.

**Comment by pmeenan** [comment]
The [server has an API](https://sites.google.com/a/webpagetest.org/docs/advanced-features/webpagetest-restful-apis) so no clicking should be needed.  You'd still need to build something that started the server and agent, ran the test using the server API, waited for the result and then shut everything down.

**Comment by iDVB** [comment]
@pmeenan ah.... API meaning REST.... there is no CLI or commands one can run in the agent or server to execute that test?

**Comment by pmeenan** [comment]
Directly, yes.  There is a [npm module](https://github.com/marcelduran/webpagetest-api) that provides a command-line interface to the API (and integrates with test specs).

**Comment by iDVB** [comment]
If so, I’m also assuming it’s possible to create an image that includes and
runs both in the same container and includes the NPM module for running the
tests. Just be a bit of fun to setup.

**Comment by pmeenan** [comment]
Yes.  The agent only knows how to poll the server for work and the server does all of the test management (and calculation for some of the metrics).  At some point it might be possible to extend the agent to run on-demand tests stand-alone but it hasn't been a priority.

**Comment by pmeenan** [comment]
Yep.  SHouldn't bee too hard to set up and configure since there are scripts to automate the setup of both the [server](https://github.com/WPO-Foundation/wptserver-install) and [agent](https://github.com/WPO-Foundation/wptagent-install) (or just re-purpose the existing docker files).

**Comment by iDVB** [comment]
🙏 I’ll post back with my Dockerfile if I get something to work.

**Comment by iDVB** [comment]
@pmeenan just dove into seeing if I can `FROM` the server and cut paste in  the agent dockerfile.... 
first off..... 
The server uses `debian:stretch-slim` as a base image
while agent uses `ubuntu:18.04`.

So I think a number of the scripts that agent uses are going to not work. :(
I'm fairly competent with Docker but Ive no experience which either of these applications. Are you able to offer any guidance on what I might try? In order to get them working in a single image/container?

**Comment by pmeenan** [comment]
The server should be able to use ubuntu:18.04 as well.  I don't think any of the packages would need to change (though the php-fpm config might need some tweaks for php version).

**Comment by iDVB** [comment]
@pmeenan I found another way to keep things simple to do it.
This way I can launch a composition in my pipeline... similar to `docker-compose` and then have the CLI act on those/that.

Here is the `docker-compose` equivant of what I'm trying to do.
https://gist.github.com/iDVB/ddb2f637ff9a0b3452976653bca09f85

Server and Agent seem to spin up ok.... but the Agent does not seem to be able to reach the server.
Any idea why?

The Agent keeps consoling over and over...
`19:10:37.638 - Get Work Error: None`

**Comment by pmeenan** [comment]
The get work should have a URL it is polling for work.  Looks like the command-line for the agent isn't including the url to reach the server.

It should look like this:
```
12:32:35.255 - Checking for work: http://agent.webpagetest.org/work/getwork.php?f=json&shards=1&reboot=1&location=Test&pc=pmeenan-l&key=xxxxxxx&version=180911.210918&screenwidth=1920&screenheight=1200&freedisk=201.162&upminutes=1604
12:32:35.261 - http://agent.webpagetest.org:80 "GET /work/getwork.php?f=json&shards=1&reboot=1&location=Test&pc=pmeenan-l&key=12345-54321&version=180911.210918&screenwidth=1920&screenheight=1200&freedisk=201.162&upminutes=1604 HTTP/1.1" 200 None
```

It usually helps to pass -vvvv as a command-line option to the agents which will provide a lot more logging.

**Comment by iDVB** [comment]
Updated gist with logs too
https://gist.github.com/iDVB/ddb2f637ff9a0b3452976653bca09f85

**Comment by pmeenan** [comment]
Does http://server:4000/ actually resolve to the server's docker container (and is it reachable from the agent container)?  If it's all local, should it be http://localhost:4000/ instead?  That's assuming 4000 is mapped to 80 for the server container.

Sorry, my docker foo is very weak so hopefully others can chime in.

**Comment by iDVB** [comment]
@pmeenan ya.... its odd... because usually with docker-compose you can use `depends_on` to map the yml node-name to Env vars that resolve to their internal addresses.

https://stackoverflow.com/questions/35832095/difference-between-links-and-depends-on-in-docker-compose-yml

I'm using this same technique on many other projects successfully.

**Comment by iDVB** [comment]
Just a thought..... does the Server ever sniff the ip that the agent is sending from and try to call it back? If so.. that would likely fail in this config since the server cannot access the agent.

**Comment by pmeenan** [comment]
No. The server is passive and the agent always connects to it.

________________________________

**Comment by iDVB** [comment]
got it working.... seems that the images don't support port mapping.... but this works...
https://gist.github.com/iDVB/ddb2f637ff9a0b3452976653bca09f85

**Comment by iDVB** [comment]
@pmeenan maybe someone should look into seeing why this container/image does not support port mapping?
